README.md
# bk1

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

    $ createdb bk1
    $ lein run -m bk1.db
    $ lein ring server-headless

    http://localhost:3002/graphql?query={transactions(account_id:1){id+amount+created_at}}


## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-03-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-03-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/bk1/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/bk1/compare/0.1.0...0.1.1


test/bk1/core_test.clj
(ns bk1.core-test
  (:require [clojure.test :refer :all]
            [bk1.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject bk1 "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [graphql-clj "0.1.20" :exclusions [org.clojure/clojure]]
                 [ring "1.5.1"]
                 [ring/ring-json "0.4.0"]
                 [ring/ring-defaults "0.2.3"]
                 [ring-cors "0.1.9"]
                 [compojure "1.5.2"]
                 [org.clojure/core.match "0.3.0-alpha4"]
                 [clojure-future-spec "1.9.0-alpha14"]
                 [postgresql "9.3-1102.jdbc41"]
                 [tentacles "0.5.1"]
                 [org.clojure/java.jdbc "0.6.1"]
                 ; https://github.com/barend/java-iban.git
                 [nl.garvelink.oss/iban "1.5.0"]]
  :main ^:skip-aot bk1.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}}
  :plugins [[lein-ring "0.9.7"]
            [cider/cider-nrepl "0.13.0-SNAPSHOT"]]
  :ring {:handler bk1.handler/app
         :auto-reload? true
         :port 3002
         :nrepl {:start? true}})


src/bk1/db.clj
(ns bk1.db
  (:require [clojure.java.jdbc :as sql]
            [environ.core :as env]
                                        ;            [tentacles.repos :as repos])
            )
  (:import (java.util UUID))
  (:refer-clojure :exclude [find]))

(def db-spec {:connection-uri (env/env :database-url "jdbc:postgresql://localhost:5432/bk1")})

(defn initial-schema [db]
  (sql/db-do-commands db
                      [(sql/create-table-ddl "entities"
                                             [[:id :serial "PRIMARY KEY"]
                                              [:name :varchar]])
                       (sql/create-table-ddl  "accounts"
                                              [[:id :serial "PRIMARY KEY"]
                                               [:name :varchar "NOT NULL"]
                                               [:iban :varchar]
                                               [:entity_id :integer "NOT NULL REFERENCES entities(id)"]
                                               [:created_at :timestamp "NOT NULL" "DEFAULT CURRENT_TIMESTAMP"]])
                       (sql/create-table-ddl "transactions"
                                             [[:id :serial "PRIMARY KEY"]
                                              [:particulars :varchar]
                                              [:code :varchar]
                                              [:reference :varchar]
                                              [:amount :numeric "NOT NULL"]
                                              [:credit :integer "NOT NULL REFERENCES accounts(id)"]
                                              [:debit :integer "NOT NULL REFERENCES accounts(id)"]
                                              [:created_at :timestamp "NOT NULL" "DEFAULT CURRENT_TIMESTAMP"]])
                       (sql/create-table-ddl "users"
                                             [[:id :serial "PRIMARY KEY"]
                                              [:name :varchar "NOT NULL"]
                                              [:entity_id :integer "NOT NULL" "REFERENCES entities(id)"]])]))

(defn add-dummy-data [db]
  (sql/insert-multi! db :entities
                     [{:id 1 :name "Entity One"}])
  (sql/insert-multi! db :accounts
                     [{:id 1 :name "Test One" :entity_id 1}
                      {:id 2 :name "Test Two" :entity_id 1}])
  (sql/insert-multi! db :transactions
                     [{:amount 42.95 :credit 2 :debit 1}
                      {:amount 12.34 :credit 2 :debit 1}]))

(defn run-and-record [db migration]
  (println "Running migration:" (:name (meta migration)))
  (migration db)
  (sql/insert! db "migrations" [:name :created_at]
                     [(str (:name (meta migration)))
                      (java.sql.Timestamp. (System/currentTimeMillis))]))

(defn migrate [& migrations]
  (sql/with-db-connection [db db-spec]
    (try (sql/db-do-commands db
          (sql/create-table-ddl "migrations"
                                [[:name :varchar "NOT NULL"]
                                 [:created_at :timestamp "NOT NULL" "DEFAULT CURRENT_TIMESTAMP"]]))
         (catch Exception _))

    (sql/db-do-commands db
                        (let [has-run? (let [result (sql/query db ["SELECT name FROM migrations"])]
                                         (set (map :name result)))]
                          (doseq [m migrations
                                  :when (not (has-run? (str (:name (meta m)))))]
                            (run-and-record db m))))))

(defn -main []
  (migrate #'initial-schema
           #'add-dummy-data))


src/bk1/handler.clj
(ns bk1.handler
  (:require [compojure.core :refer [GET POST defroutes]]
            [compojure.route :as route]
            [ring.util.response :as response]
            [ring.middleware.json :refer [wrap-json-response]]
            [ring.middleware.json :refer [wrap-json-params]]
            [ring.middleware.cors :refer [wrap-cors]]
            [ring.middleware.defaults :refer :all]
            [cheshire.core :as json]
            [bk1.graphql :as graphql]
            [clojure.walk]))

(defroutes routes
  (GET "/" [] "Hello, World!")
  (GET "/graphql" [schema query variables :as request]
       (response/response
        (graphql/execute query variables)))
  (POST "/graphql" [schema query variables :as request]
        (response/response
         (try
           (graphql/execute query (clojure.walk/stringify-keys variables))
           (catch Throwable e
             (println e)))))
  (route/resources "/" {:root ""})
  (route/not-found "Page not found"))

(def app
  (-> routes
      wrap-json-response
      (wrap-cors :access-control-allow-origin [#"http://localhost:8080" #"http://.*"]
                 :access-control-allow-methods [:get :put :post :delete])
      (wrap-defaults api-defaults)
      (wrap-json-params)))


src/bk1/graphql.clj
(ns bk1.graphql
  (:require [graphql-clj.parser :as parser]
            [graphql-clj.type :as type]
            [graphql-clj.resolver :as resolver]
            [graphql-clj.executor :as executor]
            [graphql-clj.validator :as validator]
            [graphql-clj.introspection :as introspection]
            [clojure.core.match :as match]
            [clojure.java.jdbc :as sql]
            [environ.core :as env]))

(def db-spec {:connection-uri  (env/env :database-url "jdbc:postgresql://localhost:5432/bk1")})

; http://localhost:3002/graphql?query={transactions(account_id:1){id+amount+created_at}}
; http://localhost:3002/graphql?query={account(id:1){id+name+transactions+{id+amount+created_at}}}

(def bk1-schema "
type Account {
  id: Int!
  name: String
  iban: String
  transactions: [Transaction]
}

type Transaction {
  id: Int!
  particulars: String
  code: String
  reference: String
  amount: Int
  created_at: String
}

type Query {
  transactions(account_id: Int!): [Transaction]
  transaction(id: Int!): Transaction
  accounts: [Account]
  account(id: Int!): Account
}

type Mutation {

}

schema {
  query: Query
  mutation: Mutation
}
")

(defn get-transactions [account-id]
  (sql/with-db-connection [db db-spec]
    (try (sql/query db
                    ["SELECT * FROM transactions WHERE credit = ? OR debit = ?"
                     account-id account-id])
         (catch Exception _))))

(defn get-account [account-id]
  (sql/with-db-connection [db db-spec]
    (try
      (let [account (sql/get-by-id db :accounts account-id :id)]
        (if account
          (assoc account :transactions
                 (sql/query db
                            ["SELECT * FROM transactions WHERE credit = ? or debit = ?" account-id account-id]))))
      (catch Exception e
        (println e)))))

(defn starter-resolver-fn [type-name field-name]
  (match/match
   [type-name field-name]
   ["Query" "transactions"] (fn [context parent args]
                              (get-transactions (get args "account_id")))
   ["Query" "account"] (fn [context parent args]
                         (get-account (get args "id")))
   :else nil))

(def parsed-schema (parser/parse bk1-schema))

(defn execute
  [query variables]
  (println query variables)
  (let [type-schema (validator/validate-schema parsed-schema)
        context nil]
    (executor/execute context type-schema starter-resolver-fn query variables)))
